# Lecture 14 — NumPy Arrays: A Year of Nasvhille Weather
### EES 3350/5350 · Python in Earth Science · Wednesday, October 07, 2026

In Labs 04 and 05 you worked with Nasville's **climate normals**: 30-year monthly averages. But nobody measures a monthly average. Someone measures a high, a low, and a rain gauge **every single day**, and the normals are built from decades of those days.

Today we start with **one week** of those days, typed in by hand, and use it to learn what a NumPy **array** is. Then we scale up to **all 365 days of 2025**, recorded at the Nashville airport, and answer a simple question: **was 2025 a normal year*?

**Learning objectives for today**
- import NumPy and turn a list into an **array**
- describe an array with `.shape`, `.size`, and `.dtype`
- index and slice arrays (same rules as lists!)
- do math on **every element at once** (element-wise math), and know where lists behave differently
- use NumPy's math functions (`np.exp()`) instead of the `math` module's
- make new arrays with `np.arange()` and `np.linspace()`
- load a whole data file into arrays, and boil an array down to one number with `.sum()`, `.mean()`, `.max()`, and `.min()`

---
## Part 0 : get today's files with `git pull`

This is the first time you're pulling new course material yourself. Open a **terminal** in ExploreHub and run:
```
cd ~/course-materials
git pull
ls week07
```

You should see a new `week07` folder with **two** files in it: this notebook and `nashville_daily_2025.csv` (which we'll need in Part 5).

> If this doesn't happen, you might see an error that says `error: your local changes to the following files would be overwritten by merge: week06/13_loop_patterns.ipynb. Please commit your changes or stash them before you merge.`
> Remember that Git won't overwrite your work without asking. TO keep your work **and** get the new files. Ask Dr. Morey for help.

**Before you start working don't forget to make a copy of this file, move it to the right location, and rename it appropriately**. You can do that the same way you always have OR, you can use the terminal:

```
cp week07/14_numpy_arrays.ipynb ~/LastName_EES3350/lectures/14_numpy_arrays_LastName.ipynb
ls ~/LastName_EES3350/lectures
```

---
## Part 1 : what is an array?

Here are the daily highs ($^\circ$F) at the Nashville airport for the first week of 2025 (Jan 1-7) in a list:

In [1]:
highs_list = [44, 49, 44, 43, 56, 56, 32]

print(highs_list)
print(type(highs_list))

[44, 49, 44, 43, 56, 56, 32]
<class 'list'>


*Data for this whole notebook: NOAA National Centers for Environmental Information (NCEI), Global Historical Climatology Network – Daily (GHCN-Daily), station USW00013897 (Nashville Intl AP, TN), January 1 – December 31, 2025, retrieved October 6, 2026 via the [NCEI Access Data Service](https://www.ncei.noaa.gov/support/access-data-service-api-user-documentation).*

To turn this into an array, we need NumPy. Importing it looks like the `import math` from Week 2, but with a nickname.**Everyone** abbreviates NumPy as `np`. You'll see this in every tutorial and paper.

In [2]:
import numpy as np

Now let's use `np.array()` to turn our list into an array:

In [3]:
week_highs = np.array(highs_list)

print(week_highs)
print(type(week_highs))

[44 49 44 43 56 56 32]
<class 'numpy.ndarray'>


#### An array is like a list. 
It prints without commas and its type is `numpy.ndarray`: an "n-dimensional array". That's NumPy's name for an array.
##### There are two big differences between arrays and lists:
1. in an array, every element ahs the same type
2. in an array, math happens to every element at once

Arrays can tell you about themselves:

In [4]:
print(len(week_highs))
print(week_highs.shape)
print(week_highs.size)
print(week_highs.dtype)

7
(7,)
7
int64


#### What just happened?
|**what you type**|**what it tells you**|**for `week_highs`**|
|:--|:--|:--|
|`len(week_highs)`|how many elements (same as for lists)| 7|
|`week_highs.shape`|the size along each dimension (as a tuple)| `(7,)`: one dimension, 7 long|
|`week_highs.size`|total number of elements| 7|
|`week_highs.dtype`|the **one** data type every element shares| `int64`: whole numbers|

Notice that `.shape`, `.size`, and `.dtype` have **no parentheses**. They are **attributes**: facts the array already knows about itself, not functions that have to run. `len()` is a regular function, so it does get parentheses.

#### "Every element ahs the same type": seeing it
A list can hold amix of types. An array can't, so NumPy **converts** everything to one type. Predict what each line prints, then run it:

In [6]:
print(np.array([44, 49.5]))
print(np.array([44, 49.5]).dtype)
print(np.array([44, 49.5, "hot"]))

[44.  49.5]
float64
['44' '49.5' 'hot']


Mixing a whole number with a decimal turns **everything** into decimals (`float64`): note the `44.`. Add a string and **everything** becomes a string, `'44'` included. YOu can't do math on `'44'`, so keep your arrays to one kind of data.

#### Indexing and slicing: same rules as lists
January 1 is position 0, so January 7 is position 6. Slices are `[start:stop]` with the stop **not included**.

In [7]:
print(week_highs[0])
print(week_highs[-1])
print(week_highs[3:5])

44
32
[43 56]


Look at that last day: January 7's high of 32$^\circ$F was 24 degrees colder than the day before.

<div class="alert alert-info">

### Exercise 1 : your own array

The **daily lows** for January 1-7, 2025 were 28, 25, 30, 29, 34, 27, and 27 $^\circ$F.
1. Make an array called `week_lows` from those values. Print it, its `.shape` and its `.dtype`.
2. Print the low on **January 5**.
3. Slice out the lows for **January 1-3**
    
</div>

In [10]:
# your code here
# 1

# 2

# 3


<details>
<summary style="color:green;"><strong>Answer.</strong></summary>

```python
# 1
week_lows = np.array([28, 25, 30, 29, 34, 27, 27])
print(week_lows)
print(week_lows.shape)
print(week_lows.dtype)

# 2
```
---
```
[28 25 30 29 34 27 27]
(7,)
int64
34
[28 25 30]
```

---
## Part 2 : the same math on every element
Let's convert the week's highs to Celcius. You already know how to do this with loop and `.append()`:

In [13]:
highs_c_list = []

for t in highs_list:
    highs_c_list.append((t-32)*(5/9))

print(high_c_list)

[6.666666666666667, 9.444444444444445, 6.666666666666667, 6.111111111111112, 13.333333333333334, 13.333333333333334, 0.0]


That work, and you can write it from scratch: that's three weeks of loop practice paying off. Now here's the NumPy version:

In [14]:
week_highs_c = (week_highs - 32) * (5/9)

print(week_highs_c)

[ 6.66666667  9.44444444  6.66666667  6.11111111 13.33333333 13.33333333
  0.        ]


#### What just happened?
NumPy ran the loop **for you**. When you do math between an array and a single number, NumPy applies to **every element**, in order, and hands back a **new array** the same length. This is called **element-wise**math (you'll also hear **vectorized**).
|**position**| **`week_highs[i]` ($^\circ$F)**|**`(week_highs[i] - 32) * (5/9)`**|**`week_highs_c[i]`**|
|--:|--:|:--:|--:|
|0|44|(44 - 32) * 5/9| 6.67|
|1|49|(49 - 32) * 5/9| 9.44|
|1|44|(44 - 32) * 5/9| 6.67|
|..|..|..| 6.67|
|6|32|(32 - 32) * 5/9| 0.00|

It's the same loop you wrote above. It just runs inside NumPy, where it's shorter to type and *much* faster. `week_highs` itself doesn't change. This works for `+`, `-`, `*`, `/`, and `**`.

**Your own functions work on arrays too**, as long as they only do arithmetic inside:

In [16]:
def f_to_c(temp_f):
    """Convert a temperature (or an array of temperatures) from Fahrenheit to Celsius."""
    return (temp_f - 32) * (5/9)

print(f_to_c(212))
print(f_to_c(week_highs))

100.0
[ 6.66666667  9.44444444  6.66666667  6.11111111 13.33333333 13.33333333
  0.        ]


#### Careful: lists do *not* work this way
Before you run the next two cells, **predict** what each line will print.

In [18]:
print(highs_list * 2)
print(highs_list + highs_list)

[44, 49, 44, 43, 56, 56, 32, 44, 49, 44, 43, 56, 56, 32]
[44, 49, 44, 43, 56, 56, 32, 44, 49, 44, 43, 56, 56, 32]


In [19]:
highs_list - 32

TypeError: unsupported operand type(s) for -: 'list' and 'int'

Lists treat `*` and `+` as *list* operations: `*2` **repeats the list and `+` **sticks two lists together**. And subtracting a number from a list doesn't mean anything at all, so Python raises a `TypeError`. Read the last line of the error: *unsupported operand type(s) for -: `list` and `int`.

The same symbols on the array mean math:

In [20]:
print(week_highs * 2)
print(week_highs + week_highs)

[ 88  98  88  86 112 112  64]
[ 88  98  88  86 112 112  64]


|**expression**|**if it's a list**|**if it's an array**|
|:-|:-|:-|
|`x * 2`| repeats the list (14 elements)| doubles every element (7 elements)|
|`x + x`| joins the lists end to end (14 elements)| adds element by element (7 elements)|
|`x - 32`| `TypeError`| subtracts 32 from every element|


<div class="alert alert-danger">

**Common Error:** doing math on a list when you meant an array. If you get a `TypeError` about `'list'`, or a result that's twice as long as you expedcted, check `type()` and wrape your list in `np.array()`
    
</div>

#### Two arrays at once
When **both** sides are arrays of the **same length**, NumPy pairs them up **by position**, just like `zip()` did for your parallel lists. NOAA computes a day's **average temperature** as the average of it's high and low (Arguez et al., 2012). You made `week_lows` in Exercise 1, so:

In [21]:
week_lows = np.array([28, 25, 30, 29, 34, 27, 27])

week_mean = (week_highs + week_lows) /2

print(week_highs)
print(week_lows)
print(week_mean)

[44 49 44 43 56 56 32]
[28 25 30 29 34 27 27]
[36.  37.  37.  36.  45.  41.5 29.5]


Position 0 of `week_highs` goes with position 0 of `week_lows`, position 1 with position 1, and so on. No index variable, no `range(len(...))`

<div class="alert alert-info">

## Exercise 2 : the second week of January

Here is January 8-14, 2025. 

|date|Jan 8|Jan 9|Jan 10|Jan 11|Jan 12|Jan 13|Jan 14|
|:-|-:|-:|-:|-:|-:|-:|-:|
|high (°F)|34|36|33|35|46|44|50|
|low (°F)|20|17|29|24|20|25|18|
|snowfall (in)|0.0|0.0|3.9|0.0|0.0|0.0|0.0|

1. make arrays `wk2_highs`, `wk2_lows`, and `w2_snow`.
2. use your `f_to_c()` function to make `wk2_mean_c`, each day's average temperature in C.
3. there are exactly **2.54 cm in an inch**. Make `wk2_snow_cm` and print the snowfall on January 10 cm.
4. **predict first, then run**: what happens if you try `week_highs + wk2_highs[0:5]`? Read the error message. What do you think the numbers in parentheses mean?
    
</div>

In [24]:
# your answer here
# 1

# 2

# 3

# 4


<details>
<summary style="color:green;"><strong>Answer.</strong></summary>

```python
# 1
wk2_highs = np.array([34,36,33,35,46,44,50])
wk2_lows = np.array([20,17,29,24,20,25,18])
wk2_snow = np.array([0.0,0.0,3.9,0.0,0.0,0.0,0.0])

# 2
wk2_mean_c = f_to_c((wk2_highs + week_lows) / 2)
print(wk2_mean_c)

# 3
wk2_snow_cm = wk2_snow * 2.54
print(f"Jan 10 snowfall: {wk2_snow_cm[2]:.1f} cm")

# 4
week_highs + wk2_highs[0:5]
```
---
```
[-0.55555556 -0.83333333 -0.27777778  0.          4.44444444  1.94444444
  3.61111111]
Jan 10 snowfall: 9.9 cm
ValueError: operands could not be broadcast together with shapes (7,) (5,) 
```
</details>

---
## Part 3 : `math` vs `np` - functions that work on arrays
Warm air can hold more water vapor than cold air. The **saturation vapor pressure**, $e_s$ is the most water vapor the air can hold before it starts to condense, and it depends only on temperature. A widely used formulat (Bolton, 2980) is:
$$e_s = 6.112 \, \exp\left(\frac{17.67\,T}{T + 243.5}\right)$$
where $T$ is temperature in $^\circ$C and $e_s$ is in hectopascals (hPa). the $\exp()$ means $e$ raised to that power. You've used `math.exp()` before, so let's write it as a function:

In [25]:
import math

def sat_vapor_pressure(temp_c):
    """Saturation vapor pressure (hPa) at temperature temp_c (c), from Bolton (1980)."""
    return 6.112 * math.exp(17.67 * temp_c / (temp_c + 243.5))

print(sat_vapor_pressure(20))

23.36947123406443


That works for one temperature. Now let's hand it the whole week of highs:

In [26]:
sat_vapor_pressure(week_highs_c)

TypeError: only 0-dimensional arrays can be converted to Python scalars

#### Reading the error
Start at the **bottom**: it's a `TypeError`, and the message says that only a single number (a "0-dimensional array") can be converted to a Python scalar. Now look up at the arrow: the problem is inside `math.exp()`.

The `math` module was written for **one number at a time**. It has no diea what to do with seven of them.

The fix: **NumPy has its own versions** of the math functions, and the work element-wise.
|`math` module (one number)|NumPy (one number *or* a whole array)|
|:-|:-|
|`math.exp(x)`|`np.exp(x)`|
|`math.sqrt(x)`|`np.sqrt(x)`|
|`math.log(x)`|`np.log(x)`|
|`math.sin(x)`|`np.sin(x)`|
|`math.pi`|`np.pi`|


Swap one word and redefine the function:

In [27]:
def sat_vapor_pressure(temp_c):
    """Saturation vapor pressure (hPa) at temperature temp_c (c), from Bolton (1980)."""
    return 6.112 * np.exp(17.67 * temp_c / (temp_c + 243.5))

sat_vapor_pressure(week_highs_c)

array([ 9.78783343, 11.82264601,  9.78783343,  9.42023776, 15.29583338,
       15.29583338,  6.112     ])

---
## Part 4 : making new arrays
So far we have built arrays from lists. NumPy can also build them from scratch:
|**what to write**|**what it makes**|**example**|**result**|
|:-|:-|:-|:-|
|`np.array(a_list)`| an array from a list| `np.array([44, 49, 44])`| `[44 49 44]`|
|`np.zeros(n)`| `n` zeros| `np.zeros(3)`| `[0. 0. 0.]`|
|`np.ones(n)`| `n` ones| `np.ones(3)`| `[1. 1. 1.]`|
|`np.arange(start, stop, step)`| like `range()`, but an array| `np.arange(1, 6)`| `[1 2 3 4 5]`|
|`np.linspace(start, stop, num)`| `num` evenly spaced values, **both ends included**| `np.linspace(0, 1, 5)`| `[0. 0.25 0.5 0.75 1.]`|

`np.arange()` follows the same rules as `range()`: the stop is **not** included. SO the day numbers for our first week are:

In [28]:
week_days = np.arange(1,8)
print(week_days)

[1 2 3 4 5 6 7]


#### Remember the 1.0 that wasn't?
In Lecture 09, `while rw < 1.0: rw +=1` printed an extra `Rw = 1.0", because 0.1 can't be stored exactly in binary and the tiny errors piled up. Here are two ways to build 0.0, 0.1, ..., 1.0 with NumPy:

In [29]:
print(np.arange(0, 1.0, 0.1))
print(np.linspace(0, 1, 11))

[0.  0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9]
[0.  0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1. ]


- `np.arange()` stops *before* 1.0, just like `range()` would. 
- `np.linspace()` asks a different question: not "what step>" but "**how many values** between these two ends?"
    - It calculates each value directly from the endpoints instead of adding 0.1 over and over so the ends come out exact.
 
<div class="alert alert-success">

**Tip**:
- **Want whole number steps?** | `np.arange()`
- **Want decimal spacing?** | `np.linspace()`

NumPy's own documentation reccomends `linspace` whenever the step isn't a whole number, for exactly the rounding reason we saw in Lecutre 09.
    
</div>

For example, Lab 03's water-table ratios, $R_w$ = 0.0, 0.2, ..., 1.0 are six values from 0 to 1

In [30]:
rw = np.linspace(0, 1, 6)
print(rw)

[0.  0.2 0.4 0.6 0.8 1. ]


<div class="alert alert-info">

### Exercise 3 : a saturation vapor pressure table

1. Use `np.linspace()` to make `temps_c`, temperatures from **-10 $^\circ$C to 40 $^\circ$, every 5$^\circ$C**. How many values is that? Count them before you type!)
2. Pass `temps_c` to `sat_vapor_pressure()` in **one** call to make `es_table`.
3. Use a `for` loop with `zip()` to print one line per temperature, like `-10 C: 2.9 hPa`.
    
</div>

In [33]:
# your answer here
# 1
temps_c = np.linspace(___, ___, ___)

# 2
es_table = sat_vapor_pressure(___)

#
for ___________- in __________________:
    print(f"{___:5.0f} C: {___:5.1f} hPa")

  -10 C:   2.9 hPa
   -5 C:   4.2 hPa
    0 C:   6.1 hPa
    5 C:   8.7 hPa
   10 C:  12.3 hPa
   15 C:  17.0 hPa
   20 C:  23.4 hPa
   25 C:  31.7 hPa
   30 C:  42.5 hPa
   35 C:  56.3 hPa
   40 C:  73.9 hPa


<details>
<summary style="color:green;"><strong>Answer.</strong></summary>


```python
# 1
temps_c = np.linspace(-10, 40, 11)

# 2
es_table = sat_vapor_pressure(temps_c)

#
for T, es in zip(temps_c, es_table):
    print(f"{T:5.0f} C: {es:5.1f} hPa")
```
---
```
  -10 C:   2.9 hPa
   -5 C:   4.2 hPa
    0 C:   6.1 hPa
    5 C:   8.7 hPa
   10 C:  12.3 hPa
   15 C:  17.0 hPa
   20 C:  23.4 hPa
   25 C:  31.7 hPa
   30 C:  42.5 hPa
   35 C:  56.3 hPa
   40 C:  73.9 hPa
```
From -10 to 40 every 5 is 50 / 5 = 10 steps, which means **11 values**. look down the column: roughly every 10$^\circ%C, $e_s$ about **doubles**. Also notice that arrays and loops work together: NumPy did the math, and the loop did the printing.

</details>

---
## Part 5 : scaling up - a whole year form a file

Typing one week was fine. But 2025 has **365** days, and each day has a high, a low, precipitation, and snowfall. That's 1,460 numbers to type, with 1,460 chances for a typo....Real data lives in **files**; let's learn to import data from them.

The `git pull` at the start of class downloaded a `.csv` file.

First, let's peak at the raw file,. We can do this using a `!` at the start of a line to run a **terminal command** from inside a notebook. `head -8` shows the first 8 lines of a file:

In [37]:
!head -8 nashville_daily_2025.csv

# Nashville International Airport (BNA), TN -- daily weather, Jan 1 to Dec 31, 2025 (365 days)
# Source: NOAA National Centers for Environmental Information (NCEI), Global Historical Climatology Network - Daily (GHCN-Daily),
#         station USW00013897, accessed 2026-10-06 via the NCEI Access Data Service (dataset=daily-summaries, units=standard).
# Columns: month, day, tmax (daily high, F), tmin (daily low, F), prcp (precipitation, in), snow (snowfall, in)
# One row per day, in order: row 0 is January 1.
1,1,44,28,0.00,0.0
1,2,49,25,0.00,0.0
1,3,44,30,0.00,0.0


The lines starting with `#` describe the data: where it came from and what each column means. After that, there is **one row per day**: month, day, high ($^\circ$F), low ($^\circ$F), precipitation (in), and snowfall (in). Look at the highs in rows 1-7: they're the same first week highs you typed in Part 1.

We're going to use `np.loadtxt()` with the name of the file as an input and something to tell it what separates each data point.
- `np.loadtxt()` reads a text file full of numbers into an array.
    - It skips the `#`lines on its own
- `delimiter=","` tells it the numbers are separated by commas

In [40]:
data = np.loadtxt("nashville_daily_2025.csv", delimiter=',')

print(data[:7])
print(data.shape)

[[ 1.    1.   44.   28.    0.    0.  ]
 [ 1.    2.   49.   25.    0.    0.  ]
 [ 1.    3.   44.   30.    0.    0.  ]
 [ 1.    4.   43.   29.    0.    0.  ]
 [ 1.    5.   56.   34.    1.19  0.  ]
 [ 1.    6.   56.   27.    0.    0.  ]
 [ 1.    7.   32.   27.    0.    0.  ]]
(365, 6)


#### What just happened?

This file is a table, so NumPy gave us aback a table: a **2D array**. Its shape `(365,6)` lists **rows first, then columns**. That's 365 rows (one per day) and 6 colums (one per variable). `data[:7]` printed the first 7 rows, which is the first week of January.

We'll do a lot more with 2D arrays later. For today, we only need one move: pulling out a single **column** as it's own 1D array.

In [41]:
tmax = data[:, 2]
tmin = data[:, 3]
prcp = data[:, 4]
snow = data[:, 5]

print(tmax[:7])
print(tmax.shape)
print(tmax.dtype)

[44. 49. 44. 43. 56. 56. 32.]
(365,)
float64


Inside the square brackets, the **row** comes first and the **column** second, separated by a comma. The `:` means *every row*. So `data[:, 2]` means "every row, column 2": all 365 daily highs

|**column**| **what it holds**|**our variable**|
|-:|:-|:-|
|0|month|*(not needed today)*|
|1|day|*(not needed today)*|
|2|daily high ($^\circ$F)|`tmax= data[:, 2]`|
|3|daily low ($^\circ$F)|`tmmin= data[:, 3]`|
|4|precipitation (in)|`prcp= data[:, 4]`|
|5|snowfall (in)|`snow= data[:, 5]`|

Columns count from 0, just like everything else in Python

`tmax` is the same first seven highs as `week_highs`, but now its shape is `(365,)`. Its dtype is `float64` instead of `int64` isntead of `int64`, because `np.loadtxt()` reads every number as a decimal. That's why they print as `44.` isntead of `44`.

**Everything from Parts 1-4 works exactly the same on 365 elements as it did on 7.** Look:

In [42]:
tmax_c = f_to_c(tmax)
tmean = (tmax + tmin) / 2

print(tmax_c.shape)
print(tmean.shape)

(365,)
(365,)


#### Finding a date

Row 0 of the file is January 1, so the index of any date is the *number of days before it*:

|**date**|**days before it**| **index**|
|--|:-|-:|
|Jan 1|none| 0|
|Feb 1|31 (all of January)| 31|
|Mar 1|31 + 28| 59|
|Jul 1|31 + 28 + 31 + 30 + 31 + 30| 181|
|Dec 25|334 + 24| 358|

In [44]:
print(f"Dec 25: high {tmax[358]} F, low {tmin[358]} F")

Dec 25: high 71.0 F, low 63.0 F


<div class="alert alert-info">

### Exercise 4 : finding days in the full year

1. Print the high and low on **July 4** (index 184).
2. Make a slice called `feb_highs` with every high in **February** (2025 is not a leap year). Prit it and its length.
3. Using a **negative** slice, print the lows for the **last 7 days** of the year.
    
</div>

In [45]:
# your answer here
# 1
print(f"July 4: high {tmax[184]} F, low {tmin[184]} F")

# 2
feb_highs = tmax[31:59]
print(feb_highs)
print(len(feb_highs))

# 3
print(tmin[-7:])

July 4: high 95.0 F, low 70.0 F
[51. 68. 74. 65. 70. 73. 52. 72. 47. 52. 44. 57. 47. 48. 68. 59. 41. 39.
 28. 24. 36. 44. 53. 65. 75. 80. 67. 71.]
28
[63. 64. 58. 53. 28. 25. 24.]


<details>
<summary style="color:green;"><strong>Answer.</strong></summary>

```python
# 1
print(f"July 4: high {tmax[184]} F, low {tmin[184]} F")

# 2
feb_highs = tmax[31:59]
print(feb_highs)
print(len(feb_highs))

# 3
print(tmin[-7:])
```
---
```
July 4: high 95.0 F, low 70.0 F
[51. 68. 74. 65. 70. 73. 52. 72. 47. 52. 44. 57. 47. 48. 68. 59. 41. 39.
 28. 24. 36. 44. 53. 65. 75. 80. 67. 71.]
28
[63. 64. 58. 53. 28. 25. 24.]
```
February starts at index 31 (after January's 31 days) and ends just before March 1 at index 59, so `[31:59]` gives 28 days.

`[-7:]` means "start 7 from the end, go to the end."

</details>

---
## Part 6 : one number for 365 - was 2025 normal?

Since Week 4, you've added things up with an accumulator: `total = 0` befoe a loop and `total += ...` inside it. Arrays have that loop built in:

In [47]:
total_precip = prcp.sum()
print(total_precip)

54.910000000000004


Those trailing digits are a consequence of the fact that decimals aren't stored exactly in binary, and adding 265 of them leaves a tiny crumb. The eal answer is 54.91 in. **Format when you print** and the crumb disappears:

In [49]:
normal_precip = 50.51

print(f"2025 total:   {total_precip:.2f} in")
print(f"Normal total: {normal_precip:.2f} in")
print(f"Difference:   {total_precip - normal_precip:+.2f} in")

2025 total:   54.91 in
Normal total: 50.51 in
Difference:   +4.40 in


*The normal comes from adding the 12 month precipitaiton normals in NOOA's 1991-202 US Climate Normals for Nashville.*

##### 2025 was about **4.4 in wetter** than normal. What about temperature?
|**what to write**|**what it gives you**|
|--|--|
|`a.sum()`| the total of every element|
|`a.mean()`| the average|
|`a.max()`| the biggest value|
|`a.min()`| the smallest value|

Each one also comes as a function: `np.sum(a)`, `np.mean(a)`, and so on do exactly the same thing.

In [50]:
normal_tmean = 60.8 # F the average of the 12 monthly 1991-2020 tmean normals for this station

print(f"Hottest high: {tmax.max()} F")
print(f"Coldest cold: {tmin.min()} F")
print(f"2025 average temperature: {tmean.mean():.1f} F (normal {normal_tmean} F)")

Hottest high: 101.0 F
Coldest cold: 8.0 F
2025 average temperature: 62.7 F (normal 60.8 F)


So 2025 ran about **2 $^\circ$F warmer** and about **4 inches wetter** than the 1991-2020 normal, at least at this one station.

<div class="alert alert-info">

### Exercise 5: more of 2025 in one line each

1. How much **snow** fell at the airport in 2025, in total? How much of it fell on January 10?
2. What was the **biggest single-day rainfall** of the year, in inches *and* in mm? (25.4 mm in an inch)
3. What was the **average daily high** for summer, June 1 thorugh August 31? June 1 is index 151 and August 31 is index 242.
4. The hottest afternoon of hte year was August 19 (index 230) and the coldest was January 20 (index 19). Use `sat_vapor_pressor()` and `tmax_c` to find how many times more water vapor the air *could* hold on August 19.
    
</div>

In [53]:
# your answer here
# 1
print(f"Total snowfall: {_____:.1f} in, {_____:.1f} in on Jan 10")

# 2
print(f"Wettest day: {_____:.2f} in = {_____ * 25.4:.1f} mm")

# 3
summer_highs = tmax[151:243]
print(f"Summer average high: {_____:.1f} F")

# 4
es_max = sat_vapor_pressure(tmax_c)
print(f"Aug 19: {_____:.1f} hPa, Jan 20: {_____:.1f} hPa, ratio {_____ / _____]:.0f}x")

Total snowfall: 6.1 in, 3.9 in on Jan 10
Wettest day: 3.86 in = 98.0 mm
Summer average high: 91.3 F
Aug 19: 67.6 hPa, Jan 20: 4.2 hPa, ratio 16x


<details>
<summary style="color:green;"><strong>Answer.</strong></summary>


```python
# your answer here
# 1
print(f"Total snowfall: {snow.sum():.1f} in, {snow[9]:.1f} in on Jan 10")

# 2
print(f"Wettest day: {prcp.max():.2f} in = {prcp.max() * 25.4:.1f} mm")

# 3
summer_highs = tmax[151:243]
print(f"Summer average high: {summer_highs.mean():.1f} F")

# 4
es_max = sat_vapor_pressure(tmax_c)
print(f"Aug 19: {es_max[230]:.1f} hPa, Jan 20: {es_max[19]:.1f} hPa, ratio {es_max[230] / es_max[19]:.0f}x")
```
---
```
Total snowfall: 6.1 in, 3.9 in on Jan 10
Wettest day: 3.86 in = 98.0 mm
Summer average high: 91.3 F
Aug 19: 67.6 hPa, Jan 20: 4.2 hPa, ratio 16x
```

For #1, the January 10 storm was most of the year's snow. F

For #3, the stop has has to be **243**, one past August 31 because the stop is never included
</details>

<div class="alert alert-custom" style="background-color:#e8daef; border:1px solid #d2b4de; color:#6c3483; padding:10px; border-radius:4px;">

### Bonus: degree days

Energy planners measure heating and cooling demand in **degree days**, measured from a base of **65 °F**. Each day's **cooling degree days** are how far its average temperature is *above* 65 °F, and zero on days at or below 65 °F. Heating degree days are the same idea, *below* 65 °F.

`np.maximum(a, 0)` compares every element of `a` to 0 and keeps whichever is bigger, so it turns negative numbers into 0. Use it with `tmean` to calculate Nashville's total **cooling** and **heating** degree days for 2025. 

**Was 2025 a heating year or a cooling year?**


In [ ]:
# your answer here

<details>
<summary style="color:green;"><strong>Answer.</strong></summary>

```python
cdd = np.maximum(tmean - 65, 0)
hdd = np.maximum(65 - tmean, 0)

print(f"Cooling degree days: {cdd.sum():.0f}")
print(f"Heating degree days: {hdd.sum():.0f}")
```
---
```
Cooling degree days: 2197
Heating degree days: 3034
```

Even in a warm year, Nashville needed more heating than cooling, measured this way. The trick is that `tmean - 65` is negative on cool days, and `np.maximum(..., 0)` replaces those negatives with zeros before we add them up.

</details>

---
## A useful table



|**the loop way (Weeks 4–6)**|**the NumPy way (today)**|
|:-|:-|
|`for t in temps:` + `.append(...)` to convert each value|`(temps - 32) * 5 / 9`|
|`for a, b in zip(highs, lows):` to pair up two lists|`(highs + lows) / 2`|
|`total = 0` + `total += x` (an accumulator)|`x.sum()`|
|`range(1, 8)`|`np.arange(1, 8)`|
|`while rw < 1.0: rw += 0.1` (and its float bug)|`np.linspace(0, 1, 11)`|
